### PR TITLE
chore: [CO-288] add support for ngnix templates override

### PR DESCRIPTION
**Pull request by keshavbhatt**
### What's new
The custom template override location that proxyConfGen will  use is set to `/opt/zextras/conf/nginx/templates_custom`

**Config generation flow:** If one or more template overrides are found in the custom templates location(`/opt/zextras/conf/nginx/templates_custom`), the proxyconfgen util will now use those templates to generate the config and will override the default ones that exist in the default template location (`/opt/zextras/conf/nginx/templates`) at the proxyxonfgen step while services start or by manual invocation of the tool.

**Warning header:** A warning header is written to all generated configuration files.

NOTE: The above flow won't work if a custom template path is specified explicitly to the proxyconfgen command using its -t flag. This ensures the backward compatibility of proxyconfgen’s behavior. 

### Changes in other repositories:

- **carbonio-nginx-conf** create template override dir  ->  https://github.com/Zextras/carbonio-nginx-conf/pull/22

- **carbonio-core-utils** zmfixperm support  ->  https://github.com/Zextras/carbonio-core-utils/pull/16

### CODE REVIEW

**Review by frisonisland — CHANGES_REQUESTED** [review]
I have to ask to add tests as proof of the wanted behavior.
Example: 
- test output path
- test pattern match

**Comment by keshavbhatt** [comment]
> * test output path

I don't get what you mean? output path is a constant what do we want to test about it? Sorry if I am not able to decode you again :crying_cat_face: 

> * test pattern match

can you be more specific? I did not add any new patterns in these changes. :smiling_face_with_tear:

**Comment by Polpetta** [comment]
I think @frisonisland wants the tests that:
* given the template and compiled it you have an expected result
* you want to test the regex and possible edge cases
* you want to check that the warning is printed when necessary and that fulfills behavior

**Comment by keshavbhatt** [comment]
> I think @frisonisland wants the tests that:
> 
> * given the template and compiled it you have an expected result

I will be more than happy to do that, but I have not touched the formatting of generated files they are the same and installed at the same location.

> * you want to test the regex and possible edge cases

No changes were made to affect the formatting. 

> * you want to check that the warning is printed when necessary and that fulfills behavior
 
warning is printed only when usingCustomTemplateOverride is set to true.

adding any of the tests mentioned above requires a call to expandTemplate method. So basically we are talking about writing the whole test suit for the proxyxconfgen utility. We will do that in another ticket. since we need these changes in 4.0.10

**Review by flanciotti — APPROVED** [review]

File: store/src/java/com/zimbra/cs/util/proxyconfgen/ProxyConfGen.java [hunk 1]
```diff
@@ -395,9 +400,16 @@ public static String getDefaultClientCertCaPath() {
...
     String configFilePath = configFile.getAbsolutePath();
 
+    final String templateFileName = templateFile.getName();
+    final String overrideTemplateFilePath =
+        OVERRIDE_TEMPLATE_DIR + File.separator + templateFileName;
+    final boolean usingCustomTemplateOverride =
+        !hasCustomTemplateLocationArg && Files.exists(Paths.get(overrideTemplateFilePath));
+    final String templateFilePath =
+        usingCustomTemplateOverride ? overrideTemplateFilePath : templateFile.getAbsolutePath();
+
```
Comment (frisonisland) [inline store/src/java/com/zimbra/cs/util/proxyconfgen/ProxyConfGen.java:412]: Hey Keshav, to answer your questions about testing: in the expandTemplate method **we added a check on custom template argument to infer the templateFilePath.** I think we can at least add a test about this behavior.
Everything else should have already been tested (file content, etc.) long time ago :wink: 

**I would suggest to create another method (e.g.: getTemplateFilePath) that returns the template directory output and test its value based on hasCustomTemplateLocationArg condition**.
Hope it's clear! :slightly_smiling_face:

**Review by frisonisland — CHANGES_REQUESTED** [review]

File: store/src/java/com/zimbra/cs/util/proxyconfgen/ProxyConfGen.java (hunk 1 above)
Comment (keshavbhatt) [reply store/src/java/com/zimbra/cs/util/proxyconfgen/ProxyConfGen.java:412]: Okay, my lord will do that.

**Review by frisonisland — APPROVED** [review]
